File: selfie-lib/src/commonMain/kotlin/com/diffplug/selfie/Lens.kt
/*
 * Copyright (C) 2023-2024 DiffPlug
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.diffplug.selfie

import kotlin.jvm.JvmStatic

/**
 * A camera transforms a subject of a specific type into a [Snapshot] (tutorial for
 * [jvm](https://selfie.dev/jvm/advanced#typed-snapshots)).
 */
fun interface Camera<Subject> {
  fun snapshot(subject: Subject): Snapshot
  fun withLens(lens: Lens): Camera<Subject> {
    val parent = this
    return object : Camera<Subject> {
      override fun snapshot(subject: Subject): Snapshot = lens.transform(parent.snapshot(subject))
    }
  }

  companion object {
    @JvmStatic
    fun <Subject> of(camera: Camera<Subject>, lens: Lens): Camera<Subject> = camera.withLens(lens)
  }
}

/**
 * A lens transforms a single [Snapshot] into a new [Snapshot], transforming / creating / removing
 * [SnapshotValue]s along the way (tutorial for [jvm](https://selfie.dev/jvm/advanced#lenses)).
 */
fun interface Lens {
  fun transform(snapshot: Snapshot): Snapshot
}

/** A function which transforms a string into either another string or null. */
fun interface StringOptionalFunction {
  fun apply(value: String): String?
  fun then(next: StringOptionalFunction): StringOptionalFunction = StringOptionalFunction { outer ->
    apply(outer)?.let { inner -> next.apply(inner) }
  }
}

/**
 * A lens which makes it easy to pipe data from one facet to another within a snapshot (tutorial for
 * [jvm](https://selfie.dev/jvm/advanced#compound-lens)).
 */
open class CompoundLens : Lens {
  private val lenses = mutableListOf<Lens>()
  fun add(lens: Lens): CompoundLens {
    lenses.add(lens)
    return this
  }
  fun replaceAll(toReplace: String, replacement: String): CompoundLens = mutateAllFacets {
    it.replace(toReplace, replacement)
  }
  fun replaceAllRegex(regex: String, replacement: String): CompoundLens = mutateAllFacets {
    it.replace(regex.toRegex(), replacement)
  }
  fun mutateAllFacets(perString: StringOptionalFunction): CompoundLens = add { snapshot ->
    Snapshot.ofEntries(
        snapshot.allEntries().mapNotNull { e ->
          if (e.value.isBinary) e
          else perString.apply(e.value.valueString())?.let { entry(e.key, SnapshotValue.of(it)) }
        })
  }
  fun setFacetFrom(target: String, source: String, function: StringOptionalFunction) =
      add { snapshot ->
        val sourceValue = snapshot.subjectOrFacetMaybe(source)
        if (sourceValue == null) snapshot
        else setFacetOf(snapshot, target, function.apply(sourceValue.valueString()))
      }
  fun mutateFacet(target: String, function: StringOptionalFunction) =
      setFacetFrom(target, target, function)
  private fun setFacetOf(snapshot: Snapshot, target: String, newValue: String?): Snapshot =
      if (newValue == null) snapshot else snapshot.plusOrReplace(target, SnapshotValue.of(newValue))
  override fun transform(snapshot: Snapshot): Snapshot {
    var current = snapshot
    lenses.forEach { current = it.transform(current) }
    return current
  }
}

File: python/selfie-lib/selfie_lib/Lens.py
import re
from abc import ABC, abstractmethod
from collections.abc import Iterator
from typing import Callable, Generic, Optional, Protocol, TypeVar, Union

from .Snapshot import Snapshot, SnapshotValue

T = TypeVar("T")


class Lens(Protocol):
    def __call__(self, snapshot: Snapshot) -> Snapshot:
        raise NotImplementedError


class CompoundLens(Lens):
    def __init__(self):
        self.lenses: list[Lens] = []

    def __call__(self, snapshot: Snapshot) -> Snapshot:
        current = snapshot
        for lens in self.lenses:
            current = lens(current)
        return current

    def add(self, lens: Lens) -> "CompoundLens":
        self.lenses.append(lens)
        return self

    def mutate_all_facets(
        self, perString: Callable[[str], Optional[str]]
    ) -> "CompoundLens":
        def _mutate_each(snapshot: Snapshot) -> Iterator[tuple[str, SnapshotValue]]:
            for entry in snapshot.items():
                if entry[1].is_binary:
                    yield entry
                else:
                    mapped = perString(entry[1].value_string())
                    if mapped is not None:
                        yield (entry[0], SnapshotValue.of(mapped))

        return self.add(lambda snapshot: Snapshot.of_items(_mutate_each(snapshot)))

    def replace_all(self, toReplace: str, replacement: str) -> "CompoundLens":
        return self.mutate_all_facets(lambda s: s.replace(toReplace, replacement))

    def replace_all_regex(
        self, pattern: Union[str, re.Pattern[str]], replacement: str
    ) -> "CompoundLens":
        return self.mutate_all_facets(lambda s: re.sub(pattern, replacement, s))

    def set_facet_from(
        self, target: str, source: str, function: Callable[[str], Optional[str]]
    ) -> "CompoundLens":
        def _set_facet_from(snapshot: Snapshot) -> Snapshot:
            source_value = snapshot.subject_or_facet_maybe(source)
            if source_value is None:
                return snapshot
            else:
                return self.__set_facet_of(
                    snapshot, target, function(source_value.value_string())
                )

        return self.add(_set_facet_from)

    def __set_facet_of(
        self, snapshot: Snapshot, target: str, new_value: Optional[str]
    ) -> Snapshot:
        if new_value is None:
            return snapshot
        else:
            return snapshot.plus_or_replace(target, SnapshotValue.of(new_value))

    def mutate_facet(
        self, target: str, function: Callable[[str], Optional[str]]
    ) -> "CompoundLens":
        return self.set_facet_from(target, target, function)


class Camera(Generic[T], ABC):
    @abstractmethod
    def snapshot(self, subject: T) -> Snapshot:
        pass

    def with_lens(self, lens: Lens) -> "Camera[T]":
        parent = self

        class WithLensCamera(Camera):
            def snapshot(self, subject: T) -> Snapshot:
                return lens(parent.snapshot(subject))

        return WithLensCamera()

    @staticmethod
    def of(lambda_func: Callable[[T], Snapshot]) -> "Camera[T]":
        class LambdaCamera(Camera):
            def snapshot(self, subject: T) -> Snapshot:
                return lambda_func(subject)

        return LambdaCamera()
